lib/HLSL/HLOperationLower.cpp

```diff
@@ -4173,79 +4150,113 @@ Value *TranslateBufLoad(ResLoadHelper &helper, HLResource::Kind RK,
       Args[DXIL::OperandIndex::kRawBufferLoadMaskOpIdx] =
-          GetRawBufferMaskForETy(MemTy, chunkSize, OP);
+          GetRawBufferMaskForETy(EltTy, chunkSize, OP);
+      // If we've loaded a chunk already, update offset to next chunk.
+      if (FirstLd != nullptr && opcode == OP::OpCode::RawBufferLoad)
+        Args[OffsetIdx] = Builder.CreateAdd( Args[OffsetIdx],
+                                             OP->GetU32Const(4 * LdSize));
+    }
+
+    Function *F = OP->GetOpFunc(opcode, EltTy);
+    Value *Ld = Builder.CreateCall(F, Args, OP::GetOpCodeName(opcode));
 
-    Value *Ld = GenerateBufLd(OP, Builder, opcode, Ty->getScalarType(), MemTy,
-                              chunkSize, Args, elts.begin() + i);
-    i += chunkSize;
+    // Extract elements from returned ResRet.
+    for (unsigned j = 0; j < chunkSize; j++, i++)
+      Elts[i] = Builder.CreateExtractValue(Ld, j);
 
     // Update status.
     UpdateStatus(Ld, helper.status, Builder, OP);
 
-    if (!firstLd)
-      firstLd = Ld;
+    if (!FirstLd)
+      FirstLd = Ld;
+  }
+  DXASSERT(FirstLd, "No loads created by TranslateBufLoad");
 
-    if (opcode == OP::OpCode::RawBufferLoad && i < numComponents) {
-      if (RK == DxilResource::Kind::RawBuffer)
-        // Raw buffers can't use offset param. Add to coord index.
-        Args[DXIL::OperandIndex::kRawBufferLoadIndexOpIdx] = Builder.CreateAdd(
-            Args[DXIL::OperandIndex::kRawBufferLoadIndexOpIdx],
-            OP->GetU32Const(4 * LdSize));
-      else
-        // Structured buffers increment the offset parameter.
-        Args[DXIL::OperandIndex::kRawBufferLoadElementOffsetOpIdx] =
-            Builder.CreateAdd(
-                Args[DXIL::OperandIndex::kRawBufferLoadElementOffsetOpIdx],
-                OP->GetU32Const(4 * LdSize));
+  // Convert loaded 32-bit integers to intended 64-bit type representation.
+  if (isTyped) {
+    Type *RegEltTy = Ty->getScalarType();
+    if (RegEltTy->isDoubleTy()) {
+      Function *makeDouble = OP->GetOpFunc(DXIL::OpCode::MakeDouble, RegEltTy);
+      Value *makeDoubleOpArg =
+        Builder.getInt32((unsigned)DXIL::OpCode::MakeDouble);
+      for (unsigned i = 0; i < NumComponents; i++) {
+        Value *lo = Elts[2 * i];
+        Value *hi = Elts[2 * i + 1];
+        Elts[i] = Builder.CreateCall(makeDouble, {makeDoubleOpArg, lo, hi});
+      }
+      EltTy = RegEltTy;
+    } else if (RegEltTy->isIntegerTy(64)) {
+      for (unsigned i = 0; i < NumComponents; i++) {
+        Value *lo = Elts[2 * i];
+        Value *hi = Elts[2 * i + 1];
+        lo = Builder.CreateZExt(lo, RegEltTy);
+        hi = Builder.CreateZExt(hi, RegEltTy);
+        hi = Builder.CreateShl(hi, 32);
+        Elts[i] = Builder.CreateOr(lo, hi);
+      }
+      EltTy = RegEltTy;
     }
   }
 
-  Value *retValNew = ScalarizeElements(MaybeVecMemTy, elts, Builder);
+  // Package elements into a vector.
+  Value *retValNew = nullptr;
+  if (!Ty->isVectorTy())
+    retValNew = Elts[0];
+  else {
```

Comment:
Asymmetric braces in if/else bug me.  Looks like the [style guide](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) agrees.

> // Use braces for the `if` block to keep it uniform with the `else` block.

Reply:
Done.